### PR TITLE
Rework the template mod

### PR DESCRIPTION
**Pull request by EnnuiL**
The whole thing was a mess, so why not improve it?

Here, I have done the following:
  - Grab the parts from Fabric Template Mod that were unnecessarily different here
  - Updated the README to be much clearer for the user (and also usable as a template for the end-user)
  - Use snake_case for mod IDs, use the common structure of x.y.z
  - Use version catalogs for managing all the dependency versions (If necessary, we could move it to root instead of the `gradle` subfolder)
  - Update Gradle to 7.4.2 (7.3 is necessary for Java 17 support, 7.4 for version catalogs)
  - Unbreak support for other IDEs

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,4 +1,27 @@
...
+*.iws
+
+# Visual Studio Code
+.settings/
+.vscode/
+bin/
+.classpath
+.project
+
+# MacOS
```
Comment (triphora) [inline .gitignore]:
```suggestion
# macOS
```

File: README.md [hunk 2]
```diff
@@ -1,12 +1,27 @@
-# quilt-example-mod
-Official Quilt example mod.
+# Quilt Example Mod
 
+The official Quilt example mod. You can use it as a template for you own mods!
```
Comment (triphora) [inline README.md]:
```suggestion
The official Quilt example mod. You can use it as a template for your own mods!
```

File: README.md [hunk 3]
```diff
@@ -1,12 +1,27 @@
...
-3. Enjoy!
+## Usage
+
+In order to use this mod as a template:
+
+1. Create a new repository from this template with `Use this Template`
+2. Clone the recently-created repo on your PC
+3. Make the necessary changes in order to make it yours:
+    - Update `gradle.properties` in order to use your Maven group and mod ID
+        - If you don't know which Maven group to use, and you are planning to host the mod's source code on GitHub, use `com.github.<Your_Username_Here>`
```
Comment (triphora) [inline README.md]:
```suggestion
        - If you don't know which Maven group to use, and you are planning to host the mod's source code on GitHub, use `io.github.<Your_Username_Here>`
```

File: README.md [hunk 4]
```diff
@@ -1,12 +1,27 @@
...
+    - Update `quilt.mod.json` in order to reflect your mod's metadata
+        - If you are planning to include a mod, don't forget to declare its dependency on it!
```
Comment (triphora) [inline README.md]:
```suggestion
        - If you are planning to include (jar-in-jar) a mod, don't forget to declare its dependency on it!
```

File: README.md [hunk 5]
```diff
@@ -1,12 +1,27 @@
...
+        - The icon provided here is a placeholder one. If you aren't able to replace it yet, you can delete it and remove the "icon" property
+    - Create a LICENSE file for this mod! If you don't know which license to use, check out [here](https://choosealicense.com/).
+        - If you use `LICENSE.md`, don't forget to update the buildscript in order to use that file name!
```
Comment (triphora) [inline README.md:18]:
```suggestion
        - If you use `LICENSE.md`, don't forget to update the buildscript in order to use that file name!
        - The GPLv3 and AGPLv3 are not valid mod licenses, so you can use almost any license except for those.
```

This is the bane of my existence as a modrinth moderator

File: src/main/java/com/example/example_mod/ExampleMod.java [hunk 6]
```diff
@@ -1,15 +1,18 @@
...
-	public static final Logger LOGGER = LogManager.getLogger("ExampleMod");
+	// This logger is used to write text to the console and the log file.
+	// It is considered best practice to use your mod id as the logger's name.
+	// That way, it's clear which mod wrote info, warnings, and errors.
+	public static final Logger LOGGER = LogManager.getLogger("Example Mod");
 
 	@Override
 	public void onInitialize(ModContainer mod) {
-		LOGGER.info( "Hello Quilt world from ExampleMod v{}!", mod.metadata().version().raw() );
+		LOGGER.info("Hello Quilted world!");
```
Comment (triphora) [inline src/main/java/com/example/example_mod/ExampleMod.java]:
```suggestion
 		LOGGER.info( "Hello Quilt world from ExampleMod v{}!", mod.metadata().version().raw() );
```
This was better as it was

File: src/main/java/com/example/example_mod/ExampleMod.java [hunk 7]
```diff
@@ -1,15 +1,18 @@
...
 
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 import org.quiltmc.loader.api.ModContainer;
 import org.quiltmc.qsl.base.api.entrypoint.ModInitializer;
 
 public class ExampleMod implements ModInitializer {
-	public static final Logger LOGGER = LogManager.getLogger("ExampleMod");
+	// This logger is used to write text to the console and the log file.
+	// It is considered best practice to use your mod id as the logger's name.
```
Comment (triphora) [inline src/main/java/com/example/example_mod/ExampleMod.java]:
```suggestion
	// It is considered best practice to use your mod name as the logger's name.
```

File: src/main/resources/quilt.mod.json [hunk 8]
```diff
@@ -1,42 +1,36 @@
...
-			"init": [
-				"com.examplemod.ExampleMod"
-			]
+			"init": "com.example.example_mod.ExampleMod"
 		},
 		"depends": [
 			{
 				"id": "quilt_loader",
-				"versions": ">=${loader_version}"
+				"versions": ">=0.16.0-"
```
Comment (triphora) [inline src/main/resources/quilt.mod.json:22]: Why did you remove the expansion of these?

**Review by triphora — CHANGES_REQUESTED** [review]

File: .gitignore (hunk 1 above)
Comment (EnnuiL) [reply .gitignore]: 🤦‍♀️ i followed a pattern too closely

File: README.md (hunk 3 above)
Comment (EnnuiL) [reply README.md]: Oh right, I mixed up which one is good and which one is bad

File: settings.gradle [hunk 9]
```diff
@@ -1,9 +1,23 @@
...
-		maven { url 'https://maven.quiltmc.org/repository/snapshot' }
-		gradlePluginPortal()
-		mavenCentral()
-	}
+    repositories {
+        maven {
+            name = 'Quilt'
+            url = 'https://maven.quiltmc.org/repository/release'
+        }
+        maven {
```
Comment (LambdAurora) [inline settings.gradle]: This should be removed.

**Review by LambdAurora — CHANGES_REQUESTED** [review]

File: src/main/resources/quilt.mod.json (hunk 8 above)
Comment (EnnuiL) [reply src/main/resources/quilt.mod.json:22]: Because it would get annoying with mods hard-depending on bugfix versions and versions that would otherwise be perfectly compatible with

File: src/main/java/com/example/example_mod/ExampleMod.java (hunk 6 above)
Comment (EnnuiL) [reply src/main/java/com/example/example_mod/ExampleMod.java]: I was worried that it would be too complex, but fair enough; The formatting is kinda broken though, so I'll change this manually

File: README.md [hunk 10]
```diff
@@ -1,12 +1,28 @@
...
 
-## To use this template
-1. Click on `Use this Template`
-2. Clone the resulted repo on your pc
-3. Enjoy!
+## Usage
+
+In order to use this mod as a template:
+
+1. Create a new repository from this template with `Use this Template`
```
Comment (triphora) [inline README.md]:
```suggestion
1. Create a new repository from this template with `Use this template`
```

File: build.gradle [hunk 11]
```diff
@@ -1,76 +1,87 @@
...
-		addLayer quiltMappings.mappings( "org.quiltmc:quilt-mappings:$minecraft_version+build.$quilt_mappings:v2" )
-//		addLayer loom.officialMojangMappings() // uncomment if you want mojmap too
-	})
-	modImplementation "org.quiltmc:quilt-loader:$loader_version"
-
-	modImplementation "org.quiltmc:qsl:$qsl_version"
+	minecraft libs.minecraft
+	mappings loom.layered {
+		addLayer quiltMappings.mappings("org.quiltmc:quilt-mappings:${libs.versions.quilt.mappings.get()}:v2")
+		// addLayer loom.officialMojangMappings() // Uncomment if you want to use Mojang Mappings as a fallback
```
Comment (triphora) [inline build.gradle]:
```suggestion
		// addLayer loom.officialMojangMappings() // Uncomment if you want to use Mojang Mappings as your primary mappings, falling back on QM for parameters and Javadocs
```

File: settings.gradle [hunk 12]
```diff
@@ -1,9 +1,23 @@
...
+            name = 'Quilt Snapshot'
+            url = 'https://maven.quiltmc.org/repository/snapshot'
+        }
+		// Currently needed for Intermediary and other temporary dependencies
+        maven {
+            name = 'Fabric'
+            url = 'https://maven.fabricmc.net/'
+        }
+		// Temporarily required until we update Loom Quiltflower
+		maven {
```
Comment (triphora) [inline settings.gradle:8]: Remove me

**Comment by triphora** [comment]
Could we maybe add an editorconfig file?

File: src/main/java/com/example/example_mod/ExampleMod.java [hunk 13]
```diff
@@ -1,15 +1,18 @@
...
-	public static final Logger LOGGER = LogManager.getLogger("ExampleMod");
+	// This logger is used to write text to the console and the log file.
+	// It is considered best practice to use your mod name as the logger's name.
+	// That way, it's clear which mod wrote info, warnings, and errors.
+	public static final Logger LOGGER = LogManager.getLogger("Example Mod");
 
 	@Override
 	public void onInitialize(ModContainer mod) {
-		LOGGER.info( "Hello Quilt world from ExampleMod v{}!", mod.metadata().version().raw() );
+		LOGGER.info("Hello Quilt world from Example Mod v{}!", mod.metadata().version().raw());
```
Comment (triphora) [inline src/main/java/com/example/example_mod/ExampleMod.java]: Actually, could we maybe make this use `mod.metadata().name().raw()`? :lil_pineapple:
```suggestion
		LOGGER.info("Hello Quilt world from {} v{}!", mod.metadata().name().raw(), mod.metadata().version().raw());
```

File: build.gradle (hunk 11 above)
Comment (EnnuiL) [reply build.gradle]: Hm, I'll have to think further about the Mojmap suggestion

**Comment by EnnuiL** [comment]
> Could we maybe add an editorconfig file?

Actually, I think having one would end being pretty useful with getting everything consistent; Sure, why not?

File: settings.gradle [hunk 14]
```diff
@@ -1,9 +1,14 @@
...
+            name = 'Quilt'
+            url = 'https://maven.quiltmc.org/repository/release'
+        }
+		// Currently needed for Intermediary and other temporary dependencies
+        maven {
+            name = 'Fabric'
+            url = 'https://maven.fabricmc.net/'
+        }
+        gradlePluginPortal()
+    }
```
Comment (triphora) [inline settings.gradle]: Spaces here!

**Review by triphora — COMMENTED** [review]
Just a couple more things from me!

File: build.gradle [hunk 15]
```diff
@@ -1,76 +1,87 @@
...
-		addLayer quiltMappings.mappings( "org.quiltmc:quilt-mappings:$minecraft_version+build.$quilt_mappings:v2" )
-//		addLayer loom.officialMojangMappings() // uncomment if you want mojmap too
-	})
-	modImplementation "org.quiltmc:quilt-loader:$loader_version"
-
-	modImplementation "org.quiltmc:qsl:$qsl_version"
+	minecraft libs.minecraft
+	mappings loom.layered {
+		addLayer quiltMappings.mappings("org.quiltmc:quilt-mappings:${libs.versions.quilt.mappings.get()}:v2")
+		// addLayer loom.officialMojangMappings() // Uncomment if you want to use Mojang Mappings as your primary mappings, falling back on QM for parameters and Javadocs
```
Comment (NoComment1105) [inline build.gradle]:
```suggestion
		// officialMojangMappings() // Uncomment if you want to use Mojang Mappings as your primary mappings, falling back on QM for parameters and Javadocs
```
 loom doesn't like you doing it that way

**Review by NoComment1105 — CHANGES_REQUESTED** [review]

File: .editorconfig [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+indent_style = tab
+
+[*.java]
+indent_style = tab
+
+[*.json]
+indent_style = space
+indent_size = 2
+
+[quilt.mod.json]
```
Comment (NoComment1105) [inline .editorconfig:20]: Why the double `json` based declarations?

**Review by triphora — APPROVED** [review]

File: .editorconfig (hunk 16 above)
Comment (EnnuiL) [reply .editorconfig:20]: QMJs are indented with tabs, while everything else (which would be Minecraft's data) is indented with 2 spaces

File: README.md [hunk 17]
```diff
@@ -1,12 +1,29 @@
...
+2. Clone the recently-created repo on your PC
+3. Make the necessary changes in order to make it yours:
+    - Update `gradle.properties` in order to use your Maven group and mod ID
+        - If you don't know which Maven group to use, and you are planning to host the mod's source code on GitHub, use `io.github.<Your_Username_Here>`
+    - Update `quilt.mod.json` in order to reflect your mod's metadata
+        - If you are planning to include (jar-in-jar) a mod, don't forget to declare its dependency on it!
+        - The icon provided here is a placeholder one. If you aren't able to replace it yet, you can delete it and remove the "icon" property
+    - Create a LICENSE file for this mod! If you don't know which license to use, check out [here](https://choosealicense.com/).
+        - If you use `LICENSE.md`, don't forget to update the buildscript in order to use that file name!
+        - On `quilt.mod.json`, don't forget to put in `"metadata"` a `"license"` property that contains the license's [SPDX identifier](https://spdx.org/licenses/).
```
Comment (triphora) [inline README.md:19]:
```suggestion
        - In `quilt.mod.json`, don't forget to put the license's [SPDX identifier](https://spdx.org/licenses/) under the `"license"` property in `"metadata"`.
```

Comment (triphora) [reply README.md:19]: Wait, why did I review this when it was already merged
